cli: add design guidelines (#2181)
edgelesssys/constellation: cli/README.md

# Constellation CLI design guide

This document aims to outline the design principals behind the CLI code and packages.
It should be consulted by developers when making changes to the code of packages in [`cli/internal/`](./internal/).

## Basic layout

[`cli/`](./) consists of 2 packages, [`cli/cmd/`](./cmd/), and [`cli/internal/`](./internal/), and a [`main.go`](./main.go), which serves as the entrypoint for the CLI.
No new packages or files should be added to this base directory.

[`cli/cmd/`](./cmd/) defines the root command of the CLI. Any changes concerning the root command, or additions like persistent flags (flags that can be accessed by all subcommands) should be added here.

> TODO: Discuss: Root command could be added to `cli/internal/cmd/`

Reviewer: since root is not intended to be reused and is application code, it seems fine in its current place

Any shared code should instead go into the global [`internal/` package](../internal/).

## The internal package

[`cli/internal/`](./internal/) implements logic of the CLI and is split into multiple distinct packages.
Each package should aim to implement a specific functionality. Multi purposes packages should be avoided.

The following packages are listed and explained in alphabetical order. New packages should be added here when they are created.

### [`cloudcmd`](./internal/cloudcmd/)

This package focuses on the interaction with the cloud provider.
It separates the cloud provider specific code from the rest of the CLI, and
provides a common interface for all cloud providers.
Exported functions defined here should be cloud provider agnostic, meaning there should be no functions for specific CSPs.
Rather the `cloudcmd` package should declare functions that take a `cloudprovier.Provider` as argument,
perform CSP specific logic, and return a universally usable result.

It is used by the [`cmd` package](#cmd) to handle creation of cloud resources and other CSP specific interactions.

> ### [`clusterid`](./internal/)

Reviewer 2: ```suggestion
### [`clusterid`](./internal/clusterid)
```

Defines the structure of the Constellation cluster ID file.
Logic in this package should be kept minimal.

### [`cmd`](./internal/cmd)

Defines the commands of the CLI.
Logic should be kept to input/output parsing whenever possible.
Any more complex code should usually be implemented in one of the other CLI packages.

The code here should be kept as cloud provider agnostic as possible.
Any CSP specific tasks should be handled by [`cloudcmd`](#cloudcmd).

All filepaths handled by the CLI code should originate from here.
Common filepaths are defined as constants in [`internal/constants`](../internal/constants/).

> To generate workspace correct filepaths for printing, use the functions from [`workspace.go`](./internal/cmd/workspace.go) // TODO: Discuss if we should move this to its own package

Reviewer: own pkg makes sense to me

### [`featureset`](./internal/featureset)

`featureset` defines feature gate constants for the CLI and should not implement any logic.

### [`helm`](./internal/helm)

Helm is used to manage Kubernetes deployments of a Constellation cluster.

> The package implements functionality to install, update, and manage Helm charts.

Reviewer 2: ```suggestion
The package implements functionality to install, update, and manage Helm charts.
The package includes cloud-provider specific implementations, as charts need to be rendered differently for each CSP.
Exported functions should be CSP agnostic and take a cloudprovider argument, like described for pkg `cloudcmd`.
```

Reviewer: I think `helm` can be CSP specific since otherwise the `cloudcmd` pkg would have to know about helm charts which is worse.

The charts themselves are embedded in a built CLI binary, and values are dynamically updated depending on configuration.

> The charts can be found in [`cli/internal/helm/charts/`](./internal/helm/charts/).

Reviewer 2: ```suggestion
The charts can be found in [`cli/internal/helm/charts/`](./internal/helm/charts/).
Values should be added in the chart's `values.yaml` files, if they are static i.e. don't depend on user input.
```

the CLI needs to interact with the Kubernetes API directly.
This functionality is implemented in the `kubernetes` package.

The package should be used for:

* Fetching status information about the cluster
* Creating, deleting, or migrating resources not managed by Helm

The package should not be used for anything that doesn't just require the Kubernetes API.
For example, Terraform and Helm actions should not be accessed by this package.

### [`libvirt`](./internal/libvirt)

Constellation uses libvirt to run clusters on local hardware.
To enable easy deployment across different Linux distributions, the CLI can deploy libvirt inside a Docker container.
The required files and Go code to start the container can be found in [`cli/internal/libvirt/`](./internal/libvirt/).

The code in this package should be kept minimal, and likely won't need to be changed unless we do a major refactoring of our QEMU/libvirt installation.

### [`terraform`](./internal/terraform)

Terraform is used to create cloud resources and IAM resources required for Constellation clusters.
The package implements functionality to create, update, delete, and view said resources.

> The Terraform files are embedded in a built CLI binary, and variables are dynamically update depending on configuration.

Reviewer 2: ```suggestion
The Terraform files are embedded in the CLI binary, and variables are dynamically updated depending on configuration.
```

The files can be found in [`cli/internal/terraform/terraform/`](./internal/terraform/terraform/).

> Functions of this package should be kept mostly cloud agnostic (There should be no "CreateAzureCluster" function),

Reviewer 2: ```suggestion
Functions of this package should be kept CSP agnostic (there should be no "CreateAzureCluster" function),
```

If you want to keep the "mostly", could you specify what that means?
